process/throttle/block.go

```diff
@@ -0,0 +1,144 @@
+package throttle
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/core/logger"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+var log = logger.DefaultLogger()
+
+const jumpAbovePercent = 90
+const jumpBelowPercent = 90
+const jumpAboveFactor = 0.5
+const jumpBelowFactor = 0.5
+
+type blockInfo struct {
+	succeed  bool
+	round    uint64
+	items    uint32
+	maxItems uint32
+}
+
+// blockSizeThrottle implements BlockSizeThrottler interface which throttle block size
+type blockSizeThrottle struct {
+	statistics   []*blockInfo
+	maxItems     uint32
+	mutThrottler sync.RWMutex
+}
+
+// NewBlockSizeThrottle creates a new blockSizeThrottle object
+func NewBlockSizeThrottle() (*blockSizeThrottle, error) {
+	bst := blockSizeThrottle{}
+	bst.statistics = make([]*blockInfo, 0)
+	bst.maxItems = process.MaxItemsInBlock
+	return &bst, nil
+}
+
+// MaxItemsToAdd gets the maximum number of items which could be added in one block, taking into consideration the
+// previous results
+func (bst *blockSizeThrottle) MaxItemsToAdd() uint32 {
+	bst.mutThrottler.RLock()
+	maxItems := bst.maxItems
+	bst.mutThrottler.RUnlock()
+
+	return maxItems
+}
+
+// Add adds the new state for last block which has been sent and updates the max items which could be added
+// in one block
+func (bst *blockSizeThrottle) Add(round uint64, items uint32) {
+	bst.mutThrottler.Lock()
+	bst.statistics = append(
+		bst.statistics,
+		&blockInfo{round: round, items: items, maxItems: bst.maxItems},
+	)
+	bst.mutThrottler.Unlock()
+}
+
+// Succeed sets the state of the last block which has been sent at the given round
+func (bst *blockSizeThrottle) Succeed(round uint64) {
+	bst.mutThrottler.Lock()
+	for i := len(bst.statistics) - 1; i > 0; i-- {
+		if bst.statistics[i].round == round {
+			bst.statistics[i].succeed = true
+			break
+		}
+	}
+	bst.mutThrottler.Unlock()
+}
+
+// ComputeMaxItems computes the max items which could be added in one block, taking into consideration the previous
+// results
+func (bst *blockSizeThrottle) ComputeMaxItems() {
+	//TODO: This algorithm is now something basic and could be improved
```

Comment:
we need a document with the specifications of this algorithm.

Reply:
a mathematical description of this class is needed, with a graphic of how the max value changes.

Reply:
Yes, I will try to make a short one, when I have some time. Anyway this is the first basic implementation, which will adapt the max items which could be added in one block, based on the last recent history and will always choose the next value as a middle distance between the last succeed and the last not succeed or vice-versa, depending of the last state of succeed action. This is good, because if the network speed/latency will be changed during the time, the node will be adapted based on the last recent history and not on some minimum/maximum situations recorded in the whole history.